compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -30,35 +32,151 @@ namespace mlir::iree_compiler::Preprocessing {
   return paddedResult;
 }
 
-static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
-                      ArrayRef<GPUMatmulShapeType> intrinsics) {
+static Value
+getExpandedValue(RewriterBase &rewriter, Location loc, Value expandSource,
+                 AffineMap &operandMap,
+                 SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  auto srcType = cast<RankedTensorType>(expandSource.getType());
+  auto srcShape = srcType.getShape();
+  SetVector<int64_t> operandDimsToExpand;
+  SmallVector<std::optional<int64_t>> operandDimToExpandSize(srcType.getRank(),
+                                                             std::nullopt);
+  // Get dims to expand from operand's POV.
+  for (auto [dimToExpand, sizeToExpand] : dimsToExpand) {
+    std::optional<int64_t> maybeDim = operandMap.getResultPosition(
+        getAffineDimExpr(dimToExpand, operandMap.getContext()));
+    if (maybeDim) {
+      operandDimsToExpand.insert(maybeDim.value());
+      operandDimToExpandSize[maybeDim.value()] = sizeToExpand;
+    }
+  }
+  if (operandDimsToExpand.empty()) {
+    return expandSource;
+  }
+
+  // Form reassocation indices and expanded shape.
+  SmallVector<ReassociationIndices> reassoc;
+  SmallVector<int64_t> expandedShape;
+  int64_t reassocOffset = 0;
+  for (int i = 0; i < srcType.getRank(); i++) {
+    if (operandDimToExpandSize[i]) {
+      expandedShape.append({srcShape[i], operandDimToExpandSize[i].value()});
+      reassoc.push_back(
+          ReassociationIndices{reassocOffset + i, reassocOffset + i + 1});
+      ++reassocOffset;
+    } else {
+      expandedShape.push_back(srcShape[i]);
+      reassoc.push_back(ReassociationIndices{reassocOffset + i});
+    }
+  }
+
+  return rewriter.create<tensor::ExpandShapeOp>(
+      loc, RankedTensorType::Builder(srcType).setShape(expandedShape),
+      expandSource, reassoc);
+}
+
+// expandMapsAndIterators expands the maps and iterators of an linalgOp at
+// dimensions specified by `dimsToExpand`. The general idea is we want to view
+// the dimensions specified in `dimsToExpand` as being splitted-up to two
+// dimensions. Which means the src dim and the new child/sub dim will be
+// contiguous.
+
+// For example if we have:
+// affine_map<(d0, d1, d2) -> (d0, d2, d1)>
+// iterator_types = ["parallel, "parallel", "reduce"]
+// dimsToExpand = [1, 2]
+
+// We will turn this into:
+// affine_map<(d0, d1_A, d1_B, d2_A, d2_B) -> (d0, d2_A, d2_B, d1_A, d1_B)>
+// iterator_types = ["parallel, "parallel", "parallel", "reduce", "reduce"]
+static void
+expandMapsAndIterators(SmallVector<AffineMap> &expandedMaps,
+                       SmallVector<utils::IteratorType> &expandedIterators,
+                       SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  int expandOffset = 0;
+  SmallVector<int64_t> dimsToExpandVec = llvm::to_vector(
+      llvm::map_range(dimsToExpand, [](auto &dim) { return dim.first; }));
```

Comment:
```suggestion
  auto dimsToExpandVec = llvm::to_vector_of<int64_t>(
      llvm::make_first_range(dimsToExpand));
```

Reply:
woahh, didn't know this exist, that's cool!